File: 202410/3.Longest Substring Without Repeating Characters.cpp
/*
 * @lc app=leetcode id=3 lang=cpp
 * @lcpr version=20001
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lcpr-template-start
using namespace std;
#include <algorithm>
#include <array>
#include <bitset>
#include <climits>
#include <deque>
#include <functional>
#include <iostream>
#include <list>
#include <queue>
#include <stack>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
// @lcpr-template-end
// @lc code=start
class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        unordered_map<char, int> window;
        int res = 0;

        // substring
        int left = 0, right = 0;

        while (right < s.size())
        {
            auto c = s[right];
            right++;
            window[c]++;

            while (window[c] > 1)
            {
                auto d = s[left];
                left++;
                window[d]--;
            }

            res = max(res, right - left);
        }
        return res;
    }
};
// @lc code=end

/*
// @lcpr case=start
// "abcabcbb"\n
// @lcpr case=end

// @lcpr case=start
// "bbbbb"\n
// @lcpr case=end

// @lcpr case=start
// "pwwkew"\n
// @lcpr case=end

 */

int main()
{
    Solution s;
    cout << s.lengthOfLongestSubstring(" ") << endl;
    cout << s.lengthOfLongestSubstring("abcabcbb") << endl;
    cout << s.lengthOfLongestSubstring("bbbbb") << endl;
    cout << s.lengthOfLongestSubstring("pwwkew") << endl;
    return 0;
}

File: 202503/3.Longest Substring Without Repeating Characters.java
/*
 * @lc app=leetcode id=3 lang=java
 * @lcpr version=30103
 *
 * [3] Longest Substring Without Repeating Characters
 */

import java.util.HashMap;
import java.util.Map;

// @lc code=start
class Solution {

    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> window = new HashMap<>();

        int left = 0, right = 0;
        int res = 0;
        while (right < s.length()) {
            char c = s.charAt(right);
            right++;
            window.put(c, window.getOrDefault(c, 0) + 1);

            while (window.get(c) > 1) {
                char d = s.charAt(left);
                left++;

                window.put(d, window.get(d) - 1);
            }
            res = Math.max(res, right - left);
        }
        return res;

    }
}
// @lc code=end

/*
// @lcpr case=start
// "abcabcbb"\n
// @lcpr case=end

// @lcpr case=start
// "bbbbb"\n
// @lcpr case=end

// @lcpr case=start
// "pwwkew"\n
// @lcpr case=end

 */
